### PR TITLE
client websocket refactoring

### PR DESCRIPTION
**Pull request by chappjc**
- altered context passing
- waitgroups for launched websocket handlers
- notify of wallet lock/unlock/close/create from Core

### CODE REVIEW

File: client/rpcserver/rpcserver.go [hunk 1]
```diff
@@ -258,14 +252,19 @@ func (s *RPCServer) Connect(ctx context.Context) (*sync.WaitGroup, error) {
 		}
 	}()
 
+	// Configure the websocket handler before starting the server.
+	s.mux.Get("/ws", func(w http.ResponseWriter, r *http.Request) {
+		s.wsServer.HandleConnect(ctx, w, r)
```
Comment (chappjc) [inline client/rpcserver/rpcserver.go:257]: Only way to inject the context into the websocket client since the r.Context() gets canceled the moment ServerHTTP returns.

File: client/rpcserver/rpcserver_test.go [hunk 2]
```diff
@@ -115,76 +113,35 @@ func (c *TCore) Withdraw(pw []byte, assetID uint32, value uint64, addr string) (
...
-	defer os.Remove(cert)
-	defer os.Remove(key)
+	defer os.RemoveAll(tempDir)
+
+	cert, key := tempDir+"/cert.cert", tempDir+"/key.key"
 	cfg := &Config{
-		Core: c,
-		Addr: fmt.Sprintf("localhost:%d", tPort),
+		Core: &TCore{},
+		Addr: "127.0.0.1:0",
```
Comment (chappjc) [inline client/rpcserver/rpcserver_test.go:130]: Many tests can be run in parallel now.

File: client/websocket/websocket.go [hunk 3]
```diff
@@ -106,13 +111,22 @@ func (s *Server) HandleConnect(w http.ResponseWriter, r *http.Request) {
...
 	}
-	go s.connect(wsConn, ip)
+
+	// Launch the handler for the upgraded connection. Shutdown will wait for
+	// these to return.
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		s.connect(ctx, wsConn, ip)
+	}()
```
Comment (chappjc) [inline client/websocket/websocket.go:121]: app will wait for websockets to finish.

File: client/websocket/websocket.go [hunk 4]
```diff
@@ -35,65 +35,70 @@ var (
...
 
-// HandleConnect handles the websocket connection request, creating a ws.Connection
-// and a connect thread.
-func (s *Server) HandleConnect(w http.ResponseWriter, r *http.Request) {
+// HandleConnect handles the websocket connection request, creating a
+// ws.Connection and a connect thread. Since the http.Request's Context is
+// canceled after ServerHTTP returns, a separate context must be provided to be
+// able to cancel the hijacked connection handler at a later time since this
+// function is not blocking.
+func (s *Server) HandleConnect(ctx context.Context, w http.ResponseWriter, r *http.Request) {
```
Comment (chappjc) [inline client/websocket/websocket.go:101]: There's no reason each request's context has to be the same, and this makes the context more request scoped, even if the caller (webserver/rpcserver) uses their main Connect contexts for each upgrade currently.

File: client/websocket/websocket.go [hunk 5]
```diff
@@ -235,13 +250,13 @@ type marketSyncer struct {
...
 // *dex.StartStopWaiter.
-func newMarketSyncer(ctx context.Context, cl *wsClient, feed *core.BookFeed, log dex.Logger) *dex.StartStopWaiter {
+func newMarketSyncer(cl *wsClient, feed *core.BookFeed, log dex.Logger) *dex.StartStopWaiter {
 	ssWaiter := dex.NewStartStopWaiter(&marketSyncer{
 		feed: feed,
 		cl:   cl,
 		log:  log,
 	})
-	ssWaiter.Start(ctx)
+	ssWaiter.Start(context.Background()) // wrapping Run with a cancel bound to Stop
```
Comment (chappjc) [inline client/websocket/websocket.go:259]: The context provided to newMarketSyncer was not needed.  The manager of this syncer (`websocker.(*Server).connect`) uses it as a `StartStopWaiter` via `Stop` and `WaitForShutdown`.

File: client/webserver/site/src/js/app.js [hunk 6]
```diff
@@ -335,6 +336,13 @@ export default class Application {
       }
       case 'feepayment':
         this.handleFeePaymentNote(note)
+      // case 'walletconfig':
+      //   if (note.severity !== ntfn.DATA || !note.wallet) break
+      //   const walletState = note.wallet
+      //   this.assets[wallet.assetID].wallet = walletState
+      //   this.walletMap[wallet.assetID] = walletState
+      //   const balances = this.main.querySelectorAll(`[data-balance-target="${walletState.assetID}"]`)
+      //   balances.forEach(el => { el.textContent = (walletState.balance.available / 1e8).toFixed(8) })
```
Comment (chappjc) [inline client/webserver/site/src/js/app.js:345]: Was considering shoe-horning this into a note, but it was way simpler just to type assert in the sending function in Go

File: client/webserver/site/src/js/wallets.js [hunk 7]
```diff
@@ -13,7 +13,6 @@ export default class WalletsPage extends BasePage {
     super()
     app = application
     this.body = body
-    this.notifiers = {}
```
Comment (chappjc) [inline client/webserver/site/src/js/wallets.js:16]: It's defined down below. https://github.com/decred/dcrdex/pull/601/commits/d3bc9cb9eb71b35de24181f1b75547c97d4dd646#diff-f8404ab8f5958f186ea0c27ebdc9f6efL137

File: client/webserver/site/src/js/app.js (hunk 6 above)
Comment (chappjc) [reply client/webserver/site/src/js/app.js:345]: Will remove this, but I wanted to point out that using the notify route was an option too, with the severity to indicate when it should reset the wallet state vs just notify the pages/bell. @buck54321

File: server/cmd/dcrdex/log.go [hunk 8]
```diff
@@ -50,61 +50,58 @@ var (
...
-	var err error
-	lm, err = dex.NewLoggerMaker(logWriter{}, defaultLogLevel)
-	if err != nil {
-		panic(err)
+	// Loggers must not be used before the log rotator has been initialized, or
+	// data races and/or nil pointer dereferences will occur.
+
+	makeLogger := func(name string) dex.Logger {
+		return dex.NewLogger(name, defaultLevel, logWriter{})
 	}
```
Comment (buck54321) [inline server/cmd/dcrdex/log.go:70]: Just noticing that we're setting these log levels to default, only to reset them in `parseAndSetDebugLevels`. Any reason we can't just move this `init` body to `parseAndSetDebugLevels`? We could then drop [this block](https://github.com/decred/dcrdex/blob/e4a9613f1f45a471026e4b0a1ce70ac0de3ed189/server/cmd/dcrdex/config.go#L204-L215), since the `LoggerMaker` takes care of it based on the `debugLevel` string. We're not using logging before a successful `loadConfig` anyway.

Comment (chappjc) [reply server/cmd/dcrdex/log.go:70]: Will try that out.  I wish I had split this logging fix out because it's getting hairy, or at least I'm having difficulty wrapping my head around it now for some reason.

Comment (chappjc) [reply server/cmd/dcrdex/log.go]: That turned out to be tricky, but I made the change.  We can't have multiple slog.Backends using the same logrotator because logrotator.Write isn't safe for concurrent access.  This prevents the `log` variable from being ready at startup.  This will almost certainly need to be reworked again as it's terribly hard to follow and full of limitations, but I really am sick of messing with the loggers.

**Comment by chappjc** [comment]
`'walletstate'` / `WalletStateNote`  note type added, and `'update_wallet'` route eliminated in favor of this note type.

Also, I noticed that `(*dexConnection).setEpoch` was always returning false and the calls to `refreshUser` were never called in two handers.  See https://github.com/decred/dcrdex/pull/601/commits/9795179a08931d44072cb11a768cf88e37a3f27b.  However, it wasn't clear to me what those calls accomplish, so please advise @buck54321.

File: client/websocket/websocket.go [hunk 9]
```diff
@@ -290,31 +295,38 @@ out:
...
+		errMsg := fmt.Sprintf("error unmarshalling marketload payload: %v", err)
 		s.log.Errorf(errMsg)
 		return msgjson.NewError(msgjson.RPCInternal, errMsg)
 	}
 
+	name, err := dex.MarketName(market.Base, market.Quote)
+	if err != nil {
+		errMsg := fmt.Sprintf("unknown market: %v", err)
+		s.log.Errorf(errMsg)
+		return msgjson.NewError(msgjson.RPCOrderBookError, errMsg)
```
Comment (buck54321) [inline client/websocket/websocket.go:313]: There's actually an error code for this. `msgjson.UnknownMarketError`.
